File: 1-two-sum/two-sum.cpp
class Solution {
public:

    // Brute Force solution : By iterating over the elements from the vector
    std::vector<int> bruteForceSolution(const std::vector<int>&nums, int target) {
        for(int i = 0; i < nums.size(); ++i) {  // O(n)
            for(int j = i+1; j < nums.size(); ++j) { // O(n)
                if(nums[i] + nums[j] == target) {
                    return {i,j};
                }
            }
        }
        return {-1,-1};
    }

    /*
        Analysis:
            Time Complexity : O(n^2)
            Space Complexity : O(1)
    */


    // Solution - 2 : By using the unordered map
    std::vector<int> optimisedSolution(const std::vector<int>&nums, int target) {
        std::unordered_map<int, int> elementWithIndex;

        for (int iElement = 0; iElement < nums.size(); ++iElement) {
            int remaining = target - nums[iElement];

            if (elementWithIndex.find(remaining) != elementWithIndex.end()) {
                return {elementWithIndex[remaining], iElement};
            }

            elementWithIndex[nums[iElement]] = iElement;
        }

        return {-1,-1};
    }

    /*
        Analysis:
        Time Complexity : O(n)
        Space Complexity : O(n)
    */
    
    vector<int> twoSum(vector<int>& nums, int target) {
        return optimisedSolution(nums, target);
    }
};

File: 1-two-sum/two-sum.java
class Solution {

    // Solution - 1 : Brute Force solution by checking the each pair without using the same element twice.
    private int[] solution1(int[] nums, int target) {
        for (int iIdx = 0; iIdx < nums.length; iIdx++) {
            for (int jIdx = 0; jIdx < nums.length; jIdx++) {

                // Check the condition
                if (nums[iIdx] + nums[jIdx] == target && iIdx != jIdx) {
                    return new int[] {iIdx, jIdx};
                }
            }
        }

        return new int[] {-1, -1};
    }

    /*
        Analysis:
            Time Complexity  : O(n^2)
            Space Coplexity  : O(1)
    */

    // Solution - 2 : Instead of starting the jIdx = 0 we can start the jIdx from iIdx + 1.
    private int[] solution2(int[] nums, int target) {

        for (int iIdx = 0; iIdx < nums.length; iIdx++) {
            for (int jIdx = iIdx + 1; jIdx < nums.length; jIdx++) {

                if (nums[iIdx] + nums[jIdx] == target) {
                    return new int[] {iIdx, jIdx};
                }
            }
        }

        // If we don't found any pair of elements where sum of two numbers is equal to the target
        return new int[] {-1, -1};
    }
    /*
        Analysis:
            Time Complexity  : O(n^2)
            Space Complexity : O(1)
    */

    public int[] twoSum(int[] nums, int target) {
        return solution2(nums, target);
    }
}